lib/Dialect/TTIR/Extensions/InlinerExtension.cpp

```diff
@@ -0,0 +1,57 @@
+// SPDX-FileCopyrightText: (c) 2024 Tenstorrent AI ULC
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "ttmlir/Dialect/TTIR/Extensions/InlinerExtension.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/IR/DialectInterface.h"
+#include "mlir/Transforms/InliningUtils.h"
+#include "ttmlir/Dialect/TTIR/IR/TTIR.h"
+
+using namespace mlir;
+using namespace mlir::tt::ttir;
+
+namespace {
+
+// This DialectInlinerInterface is identical to the one found in
+// mlir/lib/Dialect/Func/Extensions/InlinerExtension.cpp. We need
+// to define one for the TTIRDialect as well since the IR uses
+// FuncDialect for function definitions/calls, and TTIR for ops.
+// We need to legalize inlining for all TTIR ops.
+struct TTIRInlinerInterface : public DialectInlinerInterface {
+  using DialectInlinerInterface::DialectInlinerInterface;
+
+  // Everything can be inlined
+  bool isLegalToInline(Operation *call, Operation *callable,
+                       bool wouldBeCloned) const final {
+    return true;
+  }
+
+  bool isLegalToInline(Operation *, Region *, bool, IRMapping &) const final {
+    return true;
+  }
+
+  bool isLegalToInline(Region *, Region *, bool, IRMapping &) const final {
+    return true;
+  }
+
+  void handleTerminator(Operation *op, ValueRange valuesToRepl) const final {
+    // This should only ever be a func::ReturnOp
+    auto returnOp = cast<func::ReturnOp>(op);
+
+    // Replace usages of the functions result with the result operands of the
+    // return op
+    assert(returnOp.getNumOperands() == valuesToRepl.size());
+    for (const auto &it : llvm::enumerate(returnOp.getOperands())) {
+      valuesToRepl[it.index()].replaceAllUsesWith(it.value());
+    }
+  }
+};
+} // namespace
+
+void mlir::tt::ttir::registerInlinerExtension(DialectRegistry &registry) {
+  registry.addExtension<TTIRDialect>(
+      +[](MLIRContext *ctx, TTIRDialect *dialect) {
+        dialect->addInterfaces<TTIRInlinerInterface>();
+      });
```

Comment:
Does this have to be defined as an extension? Or could you just do
```
addInterfaces<TTIRInlinerInterface>();
```
at the dialect initialization.

Reply:
I could do that. But for FuncDialect I’m forced to register it. So, I thought it would be cleaner to just explicitly register them both rather than register only the FuncDialect Inliner and add the TTIRDialect Inliner in Initialize()

Reply:
I think I can also define my own Inliner for the func dialect (which wouldn't have the overloaded `handleTerminator` that creates `cf` ops) and register that instead.

Reply:
It does seem like arith dialect uses the interface approach, can we experiment with that @LPanosTT?

```c++
 namespace {                                         
  /// This class defines the interface for handling inlining for arithmetic         
  /// dialect operations.                         
  struct ArithInlinerInterface : public DialectInlinerInterface {    
    using DialectInlinerInterface::DialectInlinerInterface;    
                                                  
    /// All arithmetic dialect ops can be inlined.                              
    bool isLegalToInline(Operation *, Region *, bool, IRMapping &) const final {    
      return true;                                                              
    }                                                                           
  };                                                                            
  } // namespace 

  void arith::ArithDialect::initialize() {                                      
    ...                                                                                                                                                                                                                               
    addInterfaces<ArithInlinerInterface>();
```

Reply:
Yeah torch mlir does it this way too: https://github.com/llvm/torch-mlir/blob/8787970afed3c4e1497fb24c4fdeec179fcb61f6/lib/Dialect/Torch/IR/TorchDialect.cpp#L32

I think it seems like it can be done with only changes to `TTIRDialect.cpp`.  I'm not sure we need to register the inliner on behalf of func.

Reply:
See their InitAll for reference too https://github.com/llvm/torch-mlir/blob/8787970afed3c4e1497fb24c4fdeec179fcb61f6/lib/InitAll.cpp#L13.

It seems like all they do is:
- Implement the dialect interface
- addInterfaces
- Register func dialect inliner extension

Reply:
@nsmithtt it looks like they do register the inliner on behalf of func though
```
void mlir::torch::registerAllExtensions(mlir::DialectRegistry &registry) {
  mlir::func::registerInlinerExtension(registry);
  tensor::registerInferTypeOpInterfaceExternalModels(registry);
}
```

I can move the registration of the TTIR Inliner interface to TTIRDialect.cpp and use `addInterfaces` I've tried and it works. So I'll do that

Reply:
@nsmithtt @rpavlovicTT Should I create my own inliner for func? One that doesn't override the `handleTerminator` for `Block`? Just so that we leave the `cf` dialect entirely out of the picture.